controller/instance_manager_controller.go

```diff
@@ -296,140 +295,185 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
+			if !st.Ready {
+				isReady = false
+				break
 			}
+		}
 
-			if im.Status.CurrentState == longhorn.InstanceManagerStateRunning {
-				im.Status.IP = pod.Status.PodIP
-			}
-		default:
-			im.Status.CurrentState = longhorn.InstanceManagerStateError
+		if isReady {
+			im.Status.CurrentState = longhorn.InstanceManagerStateRunning
+			im.Status.IP = pod.Status.PodIP
+		} else {
+			im.Status.CurrentState = longhorn.InstanceManagerStateStarting
 		}
+	case v1.PodUnknown:
+		im.Status.CurrentState = longhorn.InstanceManagerStateUnknown
+	default:
+		im.Status.CurrentState = longhorn.InstanceManagerStateError
 	}
 
-	if im.Status.CurrentState == longhorn.InstanceManagerStateRunning {
-		if im.Status.APIVersion == engineapi.UnknownInstanceManagerAPIVersion {
-			if err := imc.versionUpdater(im); err != nil {
-				return err
-			}
+	return nil
+}
+
+func (imc *InstanceManagerController) handlePod(im *longhorn.InstanceManager) error {
+	if im.Status.CurrentState != longhorn.InstanceManagerStateError && im.Status.CurrentState != longhorn.InstanceManagerStateStopped {
+		return nil
+	}
+
+	if err := imc.cleanupInstanceManager(im.Name); err != nil {
+		return err
+	}
+	// The instance manager pod should be created on the preferred node only.
+	if imc.controllerID != im.Spec.NodeID {
+		return nil
+	}
+
+	// Since `spec.nodeName` is specified during the pod creation,
+	// the node cordon can not prevent the pod being launched.
+	unschedulable, err := imc.ds.IsKubeNodeUnschedulable(im.Spec.NodeID)
+	if err != nil {
+		return err
+	}
+	if unschedulable {
+		return nil
+	}
+
+	if err := imc.createInstanceManagerPod(im); err != nil {
+		if !types.ErrorAlreadyExists(errors.Cause(err)) {
+			return err
 		}
-	} else {
+	}
+	// The instance manager state will be updated in the next reconcile loop.
+
+	return nil
+}
+
+func (imc *InstanceManagerController) syncInstanceManagerAPIVersion(im *longhorn.InstanceManager) error {
+	// Avoid changing API versions when InstanceManagers are state Unknown.
+	// Then once required (in the future), the monitor could still talk with the pod and update processes in some corner cases. e.g., kubelet restart.
+	// But for now this controller will do nothing for Unknown InstanceManagers.
+	if im.Status.CurrentState != longhorn.InstanceManagerStateRunning && im.Status.CurrentState != longhorn.InstanceManagerStateUnknown {
 		im.Status.APIVersion = engineapi.UnknownInstanceManagerAPIVersion
 		im.Status.APIMinVersion = engineapi.UnknownInstanceManagerAPIVersion
+		return nil
 	}
 
-	if err := imc.syncInstanceManagerPDB(im); err != nil {
-		return err
+	if im.Status.CurrentState == longhorn.InstanceManagerStateRunning && im.Status.APIVersion == engineapi.UnknownInstanceManagerAPIVersion {
+		if err := imc.versionUpdater(im); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
+func (imc *InstanceManagerController) syncMonitor(im *longhorn.InstanceManager) error {
+	enableMonitor := false
+	disableMonitor := false
```

Comment:
NIT: combine these into one boolean `isMonitorRequired := im.Status.CurrentState == longhorn.InstanceManagerStateRunning` since they are exclusive of each other (true && true wouldn't make much sense)

Then you can just do this, with a bit of easy refactor, see my previous monitor suggestions
No need to check if isMonitoring the enable/disable should handle that internally.
```
if im.state == unknown {
    return nil
}

if isMonitorRequired {
   return enableMonitor(im)
} else {
   return disableMonitor(im)
}
```

Reply:
Make sense. I will refactor this function based on your suggestions.

Reply:
I refactored only a part of the monitor logic to fit the modification here. Further refactoring will be handled in https://github.com/longhorn/longhorn/issues/2441